Quote: Rename the align attribute to textAlign

carolinan opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
_Updated September 27 2023._

Renames the align attribute to textAlign.
Partial for https://github.com/WordPress/gutenberg/issues/34454
There may be other blocks where the attribute should be renamed so the issue can be kept open until that is confirmed and solved.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->
The block used the `align` attribute for the text alignment.
- This was inconsistent with (some) other blocks that use `align` for the block alignment and `textAlign` for text alignment.
- It conflicted with the potential addition of the align block support; see details in the issue above.

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->
-  Renames the attribute
- Migrates the current values to the new attribute name

## Testing Instructions
_Updated September 27 2023._

1. Deactivate Gutenberg and add a quote block with centered text:

```
<!-- wp:quote {"align":"center"} -->
<blockquote class="wp-block-quote has-text-align-center"><!-- wp:paragraph -->
<p>This is a quote</p>
<!-- /wp:paragraph --><cite>This is a citation</cite></blockquote>
<!-- /wp:quote -->
```
2. Apply the PR and activate Gutenberg.
3. Refresh the page where you added the quote. Confirm that there are no block validation errors and that the block looks the same as before.
4. Enter the code editor mode. Confirm that the code for the block uses "textAlign" instead of "align".
```
<!-- wp:quote {"textAlign":"center"} -->
<blockquote class="wp-block-quote has-text-align-center"><!-- wp:paragraph -->
<p>This is a quote</p>
<!-- /wp:paragraph --><cite>This is a citation</cite></blockquote>
<!-- /wp:quote -->
```

## Screenshots or screencast 
Before:
<img width="1221" alt="quote block with centered text" src="https://github.com/WordPress/gutenberg/assets/7422055/e9f78256-2df9-430a-be00-ae6fda9b0d52">

After:
<img width="1221" alt="quote block, with the pull request applied" src="https://github.com/WordPress/gutenberg/assets/7422055/af0587dc-e75d-4e8e-b993-fc5b118aebe7">

carolinan committed:
Quote: Rename align attribute to textAlign

carolinan committed:
Update deprecated.js

carolinan committed:
Update core__quote__deprecated-2.serialized.html

carolinan commented:
Can I go ahead and add the [block support for alignment ](https://github.com/WordPress/gutenberg/issues/32507)in this PR? I can do a second one line PR, but the change relies on this PR to work.

oandregal commented:
@carolinan It makes sense that we add the block support here to see the full picture.

My understanding is that the available values for (wide) align are: left, center, right, wide, full. And the available values for (text) align are: left, center, right. Given the two sets have common values but will produce different serialization, if we have an existing quote with `align: right`, how do we know whether it's text or width?

It sounds like if we want to add support for wide align to this block we may need to do it without the block support?

cc @ramonjd @andrewserong for layout thoughts.

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Add alignments for testing

carolinan commented:
I added the alignments so that we can test them.

andrewserong commented:
Thanks for the ping!

> Given the two sets have common values but will produce different serialization, if we have an existing quote with align: right, how do we know whether it's text or width?

Oh, great question. Just digging around in other blocks, and I see that the Heading block only opts in to the `wide` and `full` alignments ([here](https://github.com/WordPress/gutenberg/blob/7a6d9525434fecf9c84b46ca8f0d7d4bfef0449d/packages/block-library/src/heading/block.json#L30)), to keep them separate, and from the looks of it to avoid the collision with `left`, `right` and `center` from back when it used the `align` attribute instead of `textAlign` for the text alignment attribute. E.g. here is where the deprecation logic is handled: https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/heading/deprecated.js#L57-L64

It was a couple of years ago, but for the Quote block, it sounds like we might be able to borrow a similar technique as used by @ntsekouras in https://github.com/WordPress/gutenberg/pull/26492 to fix up the collision in the heading block?

This is all with the assumption that the overall objective is to add `wide` and `full` alignment support to the Quote block 🙂. Is that the case?

carolinan commented:
Yes the objective is both consistency across blocks and to enable the wide full / left center right alignments.

carolinan committed:
Update fixtures, add align block support to the deprecation

carolinan commented:
I might be missing / misunderstanding something important so I'm happy hand this over to anyone :)

andrewserong commented on test/integration/fixtures/blocks/core__quote__deprecated-2.serialized.html:
```diff
@@ -1,5 +1,5 @@
 <!-- wp:quote {"align":"right"} -->
-<blockquote class="wp-block-quote has-text-align-right"><!-- wp:paragraph -->
+<blockquote class="wp-block-quote alignright"><!-- wp:paragraph -->
```
This changes the output of the deprecation, which means that for a user who opens up this older markup, what once was a setting to align text to the right now sets the overall block to be aligned right.

carolinan replied:
This was not a manual change, I ran the script to update the fixtures - so the error is in the code for the deprecation, correct?

andrewserong replied:
> so the error is in the code for the deprecation, correct?

Yes, if we make an update to the deprecations code that results in having to regenerate the serialized version, then it's a good signal that we might have to make a change to the logic in the `deprecated.js` file. For this case, a good indication that the deprecation code is working correctly is if the existing `core__quote__deprecated-2.serialized.html` markup doesn't require any changes to be made in this PR. Ideally, the output of all existing deprecations should remain unchanged.

andrewserong reviewed (commented):
> and to enable the wide full / left center right alignments.

Ah, I see! Unfortunately, due to older block markup using `align` for storing the text alignment value, I don't think it will be possible to safely opt in to `left`, `center`, `right` in a backwards compatible way, due to the same issues as occurred in the Heading block. I'd recommend shifting the scope to opting in to `wide` and `full` alignments, and leave `left`, `center`, and `right` off for this block (similar to current behaviour for the Heading, Post Title, Query Title, and Site Tagline blocks), but preserve the migration from left/center/right `align` values to `textAlign` values.

From a quick look at https://github.com/WordPress/gutenberg/issues/32507, it looks like the main request was for wide alignment, so hopefully that'll still help improve the consistency for that + full width?

carolinan committed:
Remove left/right/center alignments

ntsekouras commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -34,6 +35,76 @@ export const migrateToQuoteV2 = ( attributes ) => {
+const TEXT_ALIGN_OPTIONS = [ 'left', 'right', 'center' ];
+
+// Migrate existing text alignment settings to the renamed attribute.
+const migrateTextAlign = ( attributes ) => {
```
I haven't checked the code but if we add a migration, we need to add it in all existing deprecations.

carolinan replied:
I was not able to figure out how to do this. The deprecations each use different logic for the existing `migrate` (migrateToQuoteV2), I am not sure how to combine the two migrations in `v0` and `v1`.

ntsekouras replied:
In general you combine all the migration actions into a new `migrate` function for each deprecation. This might need more eyes though because it's being called in `edit` as well. --cc @oandregal @ellatrix

aaronrobertshaw replied:
All the deprecation migrate functions now call `migrateTextAlign`, so resolving this conversation.

andrewserong commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -34,6 +35,76 @@ export const migrateToQuoteV2 = ( attributes ) => {
+			},
+		},
+	},
+	save( { attributes } ) {
+		const { textAlign, value, citation } = attributes;
+
+		const className = classnames( {
+			[ `has-text-align-${ textAlign }` ]: textAlign,
+		} );
+
+		return (
+			<blockquote { ...useBlockProps.save( { className } ) }>
+				<RichText.Content multiline value={ value } />
+				{ ! RichText.isEmpty( citation ) && (
+					<RichText.Content tagName="cite" value={ citation } />
+				) }
+			</blockquote>
+		);
+	},
```
I believe this save function should be a snapshot of the state of the save function prior to the changes in this PR. I.e. if we switch this version back to using `align` for generating `has-text-align-${ align }`, then it should hopefully get the `core__quote__deprecated-2.serialized.html` fixture passing again without that fixture needing to be changed.

carolinan committed:
Copy save function to deprecation v4. Combine migrations in deprecation v2 and v3.

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Update deprecated.js

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

ntsekouras commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -109,7 +187,7 @@ const v2 = {
 		},
 	},
-	migrate: migrateToQuoteV2,
+	migrate: compose( migrateToQuoteV2, migrateTextAlign ),
```
Nitpick: You're using this and `migrateTextAlign( migrateToQuoteV2( attributes ) );`. Can we just use the second version and not load `compose`. I mean it's just a function call.

carolinan replied:
Like this?
```
	migrate( attributes ) {
		return migrateTextAlign( migrateToQuoteV2( attributes ) );
	},
```

ntsekouras replied:
Yes

ntsekouras committed:
build docs

ntsekouras commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -34,6 +35,83 @@ export const migrateToQuoteV2 = ( attributes ) => {
+
+// Version before the 'align' attribute was replaced with 'textAlign'.
+const v4 = {
+	attributes: {
```
The `isEligible: ( { align } ) => TEXT_ALIGN_OPTIONS.includes( align ),` part is missing from this deprecation.

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

ntsekouras commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -34,6 +35,83 @@ export const migrateToQuoteV2 = ( attributes ) => {
+			</blockquote>
+		);
+	},
+	migrate: compose( migrateToQuoteV2, migrateTextAlign ),
```
In this new deprecation, we are always in `v2` so we just need to `migrateTextAlign`.

carolinan committed:
Try removing compose, add isEligible

carolinan committed:
Update deprecated.js

carolinan committed:
Update transforms.native.js.snap

ntsekouras commented on test/integration/fixtures/blocks/core__quote__deprecated-2.serialized.html:
```diff
@@ -1,5 +1,5 @@
 <!-- wp:quote {"align":"right"} -->
```
This doesn't seem right.  The final html should have the new attribute..

carolinan replied:
There is something wrong with the deprecation still.  @andrewserong suggested (above) that the fixture files should be unchanged.
I have not manually changed them, I ran the command `npm fixtures:regenerate` because of the reported test failure.

andrewserong replied:
> There is something wrong with the deprecation still.

Yes, that seems like it might be the case.

>  @andrewserong suggested (above) that the fixture files should be unchanged.

I think I might have been wrong about that, apologies, I think I might have missed some nuance in earlier reviews. In principle the block markup of this fixture should be unchanged (that is, we'd expect the HTML to be `<blockquote class="wp-block-quote has-text-align-right">`), but we'd also expect the block attributes in the block comment to have a value for `textAlign` in the serialized version.

carolinan replied:
Who is the best person to ping to review the deprecation?

aaronrobertshaw replied:
It could just be me but the adoption of the `align` block support at the same time as this renaming/deprecation makes it harder to reason about. 

That said, I agree, the results of this deprecation fixture aren't correct. As @andrewserong noted, the text align class should still be present and the attribute in the bloc comment delimiter updated to `textAlign` if it was migrated properly.

Fixing `migrateTextAlign`, or its wrapping of `migrateToQuoteV2`, gets this fixture to fail due to the incorrect block attributes.

ntsekouras commented:
@WordPress/native-mobile can you take a look here about the respective failing mobile tests? Thanks!

dcalhoun commented:
> @WordPress/native-mobile can you take a look here about the respective failing mobile tests? Thanks!

It appears this may be as simple as updating a hard-coded reference to "blocks that support alignment" in the test file. Ideally, that reference utilized each block's configuration itself rather than a static list of supported blocks. 

```diff
diff --git a/packages/components/src/mobile/utils/test/index.native.js b/packages/components/src/mobile/utils/test/index.native.js
index f24f2644b4..ad7a07be1f 100644
--- a/packages/components/src/mobile/utils/test/index.native.js
+++ b/packages/components/src/mobile/utils/test/index.native.js
@@ -76,6 +76,7 @@ describe( 'isContainerRelated', () => {
 		'core/column',
 		'core/buttons',
 		'core/button',
+		'core/quote',
 	];
 
 	beforeAll( () => {

```

There is still a chance the changes proposed in this PR disrupt or break the mobile editor, but it is not clear to me yet. 

> 3\. Refresh the page where you added the quote. Confirm that there are no block validation errors, and that the inner content, now a paragraph block, is centered.

@carolinan to clarify, is the above quote to say the paragraph block _itself_ or its _text_ is centered? Also, if you are willing to add screenshots of the expected outcome for testing instruction steps 3 and 4, that would help me verify support in the mobile editor. Thanks!

carolinan commented:
@dcalhoun I have updated the description, testing instructions and screenshots, because the PR is almost a year old. I hope it is clearer.

The text alignment option on the quote block affects the default alignment of all **text-based** inner blocks using text-align:
```
.has-text-align-center {
	text-align: center;
}
```
If the quote has text align set to center, the inner blocks need to set their own text-alignment and override the parent, if they don't want the text to be centered.
If the quote has an image block, the image is not centered, but the caption is.
>  is the above quote to say the paragraph block itself or its text is centered? 

So I guess it is the second, its text is centered.

carolinan committed:
Update index.native.js

dcalhoun commented:
> The text alignment option on the query block affects the default alignment of all **text-based** inner blocks using text-align:
> 
> ```
> .has-text-align-center {
> 	text-align: center;
> }
> ```

Thank you for clarifying. It appears the proposed changes break this functionality for the native editor. The native editor does not benefit from inherited styles, there is logic in place to "pass down" the parent alignment. 

<details><summary>Screenshots: Before vs After</summary>

| Before | After |
| - | - |
| ![quote-text-alignment-before](https://github.com/WordPress/gutenberg/assets/438664/9efabc8f-6fa1-4eae-9a5e-ea5a10520f0b) | ![quote-text-alignment-after](https://github.com/WordPress/gutenberg/assets/438664/f0f2238d-98b8-4e8d-b499-11c4002ccc63) |

</details>

@geriux fixing this regression can be done by simply renaming the attribute used to determine `parentTextAlignment` for the Paragraph block. 

<details><summary>Rename Attribute Diff</summary>

```diff
diff --git a/packages/block-editor/src/components/block-list/block-list-item.native.js b/packages/block-editor/src/components/block-list/block-list-item.native.js
index f1885196f6..cc3ae47a9d 100644
--- a/packages/block-editor/src/components/block-list/block-list-item.native.js
+++ b/packages/block-editor/src/components/block-list/block-list-item.native.js
@@ -98,7 +98,7 @@ function BlockListItemContent( {
 			const name = getBlockName( clientId );
 			const parentName = getBlockName( rootClientId );
 			const { align } = getBlockAttributes( clientId ) || {};
-			const { align: parentBlockAlign } =
+			const { textAlign: parentBlockAlign } =
 				getBlockAttributes( rootClientId ) || {};
 
 			return {
```

</details>

https://github.com/WordPress/gutenberg/blob/3cc2a900e0beddc7052c81bdcc8e540d0d7a1b04/packages/block-library/src/paragraph/edit.native.js#L47-L51

However, even though the tests pass, I'm guessing this will introduce other regressions. Does it make sense to introduce an additional `parentBlockTextAlign` prop that `BlockListItem` passes to children instead? I suppose until _all_ `align` renames are completed, we will need to check all possible sources for text alignment. WDYT?

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

dcalhoun commented:
I opened https://github.com/WordPress/gutenberg/pull/52293 to address the mobile editor shortcomings outlined in https://github.com/WordPress/gutenberg/pull/42960#issuecomment-1599474353.

dcalhoun committed:
fix: Paragraphs inherit parent text alignment (#52293)

The Paragraph `align` attribute was renamed to `textAlign`. Hence, these
changes adapt the block list to pass the relevant attribute to its child
blocks.

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

aaronrobertshaw commented on packages/block-library/src/quote/block.json:
```diff
@@ -23,11 +23,12 @@
 		}
 	},
 	"supports": {
+		"align": [ "wide", "full" ],
```
I think it could help to split out the adoption of align support into a follow-up PR to avoid confusion and draw a clear line between the renaming of the old attribute and the new one.

aaronrobertshaw commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -34,6 +34,84 @@ export const migrateToQuoteV2 = ( attributes ) => {
+	},
+	supports: {
+		anchor: true,
+		__experimentalOnEnter: true,
```
If we're aiming for parity between the current supports config and the deprecation to reduce cognitive load when revisiting these deprecations in future, maybe we should also include the `__experimentalOnMerge` flag as well?

carolinan replied:
Yes, good catch, since this PR was opened the supports have been changed.

aaronrobertshaw commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -87,7 +165,9 @@ const v3 = {
-	migrate: migrateToQuoteV2,
+	migrate( attributes ) {
+		return migrateTextAlign( migrateToQuoteV2( attributes ) );
+	},
```
```suggestion
	migrate( attributes ) {
		const [ v2Attributes, innerBlocks ] = migrateToQuoteV2( attributes );
		const v4Attributes = migrateTextAlign( v2Attributes );
		return [ v4Attributes, innerBlocks ];
	},
```

This needs to take into account the fact that `migrateToQuoteV2` will return an array containing the attributes and inner blocks. Rather than update all the places where `migrateToQuoteV2` was wrapped, it might be easier to refactor `migrateTextAlign` to accept the array with inner blocks and return them as well.

aaronrobertshaw commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -109,7 +189,9 @@ const v2 = {
 	},
-	migrate: migrateToQuoteV2,
+	migrate( attributes ) {
+		return migrateTextAlign( migrateToQuoteV2( attributes ) );
```
As noted earlier, this also needs fixing to accept the array returned by `migrateToQuoteV2`

aaronrobertshaw commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -151,15 +233,17 @@ const v1 = {
-					: 'is-style-large',
-			} );
+			return migrateTextAlign(
+				migrateToQuoteV2( {
+					...restAttributes,
+					className: attributes.className
+						? attributes.className + ' is-style-large'
+						: 'is-style-large',
+				} )
+			);
 		}
 
-		return migrateToQuoteV2( attributes );
+		return migrateTextAlign( migrateToQuoteV2( attributes ) );
```
As noted earlier, this also needs fixing to accept the array returned by `migrateToQuoteV2`

aaronrobertshaw commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -206,12 +290,14 @@ const v0 = {
-				...restAttributes,
-			} );
+			return migrateTextAlign(
+				migrateToQuoteV2( {
+					...restAttributes,
+				} )
+			);
 		}
 
-		return migrateToQuoteV2( attributes );
+		return migrateTextAlign( migrateToQuoteV2( attributes ) );
```
As noted earlier, this also needs fixing to accept the array returned by `migrateToQuoteV2`

aaronrobertshaw reviewed (commented):
Thanks for the ping to review @carolinan 👍 

It seems like this PR is close but after some testing, there appear to be a few issues with the deprecations. Some things I noticed were:

- The wrapping of calls to `migrateToQuoteV2` doesn't take into account that it can return  `[ attributes, innerBlocks ]`
- Creating a deprecated version of the Quote block with alignment on trunk before switching to this PR branch and reloading the editor, doesn't seem to complete the migration, the `align` attribute remains when viewing the markup in the code editor
- Pasting in the same deprecated block markup created above, does trigger both blocks to migrate the attribute within the code editor
- The deprecation in this PR doesn't have a dedicated fixture
- Some other past deprecations also need their own fixtures (these all help debug migration issues like we have now, as well as regressions, so are worth it in my view)
- The modified deprecation fixture for the `v2` deprecation is incorrect
- Creating a test fixture out of the latest deprecated block markup and regenerating fixture files does migrate the align attribute to textAlign

The adoption of the align block support at the same time, could be confusing when revisiting this for future deprecations. Could it be a follow-up?

Unfortunately, I'm a bit short on bandwidth to be able to dive much deeper into these deprecations at the moment. I'm happy to keep checking in though and reviewing where possible, if that helps.

After fixing up `migrateTextAlign`, I'd suggest getting some dedicated fixtures for each deprecation. That should help narrow down where any further gremlins may lie.

FYI, the block markup generated on trunk for me was:

```
<!-- wp:quote {"align":"right"} -->
<blockquote class="wp-block-quote has-text-align-right"><!-- wp:paragraph -->
<p>A wise quote.</p>
<!-- /wp:paragraph --><cite>Anon.</cite></blockquote>
<!-- /wp:quote -->
```

carolinan committed:
Fixture for version with align attribute

carolinan committed:
Revert the block support change

carolinan committed:
Update supoports in the deprecation

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Try to fix the migration

carolinan commented:
I tried updating the migration, but it caused the mobile unit tests to fail.

SiobhyB commented on packages/components/src/mobile/utils/test/index.native.js:
```diff
@@ -76,6 +76,7 @@ describe( 'isContainerRelated', () => {
 		'core/column',
 		'core/buttons',
 		'core/button',
+		'core/quote',
```
@carolinan, this list should only contain blocks that support wide alignment. You can see the relevant function the currently failing test uses to check for each of the block's alignment:

https://github.com/WordPress/gutenberg/blob/1fb1343c42a59d5062660a7da20a4bff4d06e369/packages/components/src/mobile/utils/alignments.native.js#L34-L44  

As the quote block does not support wide alignment, it should be removed from this list.

carolinan replied:
Thank you! I understand now. This alignment was supported and then I was asked to remove it and I missed this.

SiobhyB replied:
Thanks for making that change!

SiobhyB reviewed (changes requested):
@carolinan, the mobile tests should pass once the line I pointed to is removed. If you'd like further review from the native team, feel free to ping us using the `WordPress/native-mobile` @-mention :)

carolinan committed:
Tests: Remove quote from currentlySupportedBlocks in isContainerRelated

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

SiobhyB reviewed (commented):
@carolinan, I tested the changes on both mobile and the web, and can confirm I didn't spot any functional regressions. I also created a companion PR in the Gutenberg Mobile repo, to verify that the unit tests pass there: https://github.com/wordpress-mobile/gutenberg-mobile/pull/6267. 

As I don't often work with the web, and given @aaronrobertshaw's previous involvement, I'd be happy to approve after he or someone else from the web team takes another look. Thanks!

aaronrobertshaw commented on packages/block-library/src/quote/edit.js:
```diff
@@ -132,7 +132,9 @@ export default function QuoteEdit( {
-						{ ...( ! isWebPlatform ? { textAlign: align } : {} ) }
+						{ ...( ! isWebPlatform
+							? { textAlign: attributes.textAlign }
+							: {} ) }
```
```suggestion
						{ ...( ! isWebPlatform ? { textAlign } : {} ) }
```

We can make this more concise given `textAlign` has already been destructured from attributes earlier.

aaronrobertshaw commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -151,15 +240,22 @@ const v1 = {
+			const v4Attributes = migrateTextAlign( v1Attributes );
+
+			return {
+				v4Attributes,
+				innerBlocks,
 				className: attributes.className
 					? attributes.className + ' is-style-large'
 					: 'is-style-large',
-			} );
+			};
```
Probably best to stick to the normal tuple return here

aaronrobertshaw replied:
Also, the `className` attribute is migrated to v2 after attributes have been migrated to v4 already. This could cause something unexpected down the road.

aaronrobertshaw commented on packages/block-library/src/quote/deprecated.js:
```diff
@@ -109,7 +194,11 @@ const v2 = {
 	},
-	migrate: migrateToQuoteV2,
+	migrate( attributes ) {
+		const [ v2Attributes, innerBlocks ] = migrateToQuoteV2( attributes );
+		const v4Attributes = migrateTextAlign( v2Attributes );
+		return [ v4Attributes, innerBlocks ];
```
This appears to be repeated quite a lot, it would make things easier to grok if we could make this more concise.

aaronrobertshaw reviewed (commented):
Thanks for all the hard work and effort that has gone into wrangling this one @carolinan 👍 

There are still a couple of issues with the deprecations that we should probably iron out.

There's a lot of double handling of migrated attributes and inner blocks. If we made the (excluding `migrateToQuoteV2`) migration functions match the `migrate` signature, we could chain things together more. 

As you can appreciate, deprecations are hard enough to grok without making the code harder to read too.

Changing the signatures to match `migrate` would also make it easier to update all the existing deprecations when new ones are needed.

The v2 migration of the large style made me do a double take as it didn't return the normal `[attributes, innerBlocks]` tuple. While it passes, I'd suggest we tweak that too. 

Lastly, that v2 migration now migrates the classname in the attributes "out of order". The `className` attribute gets set after everything has already been migrated to v4 attributes.

<details>
<summary>Diff with suggested changes</summary>

```diff
diff --git a/packages/block-library/src/quote/deprecated.js b/packages/block-library/src/quote/deprecated.js
index a40eeb894b..fcf6776b76 100644
--- a/packages/block-library/src/quote/deprecated.js
+++ b/packages/block-library/src/quote/deprecated.js
@@ -37,13 +37,28 @@ export const migrateToQuoteV2 = ( attributes ) => {
 const TEXT_ALIGN_OPTIONS = [ 'left', 'right', 'center' ];
 
 // Migrate existing text alignment settings to the renamed attribute.
-const migrateTextAlign = ( attributes ) => {
+const migrateTextAlign = ( attributes, innerBlocks ) => {
 	const { align, ...rest } = attributes;
 	// Check if there are valid alignments stored in the old attribute
 	// and assign them to the new attribute name.
-	return TEXT_ALIGN_OPTIONS.includes( align )
+	const migratedAttributes = TEXT_ALIGN_OPTIONS.includes( align )
 		? { ...rest, textAlign: align }
 		: attributes;
+
+	return [ migratedAttributes, innerBlocks ];
+};
+
+// Migrate the v2 blocks with style === `2`;
+const migrateLargeStyle = ( attributes, innerBlocks ) => {
+	return [
+		{
+			...attributes,
+			className: attributes.className
+				? attributes.className + ' is-style-large'
+				: 'is-style-large',
+		},
+		innerBlocks,
+	];
 };
 
 // Version before the 'align' attribute was replaced with 'textAlign'.
@@ -169,9 +184,7 @@ const v3 = {
 		);
 	},
 	migrate( attributes ) {
-		const [ v3Attributes, innerBlocks ] = migrateToQuoteV2( attributes );
-		const v4Attributes = migrateTextAlign( v3Attributes );
-		return [ v4Attributes, innerBlocks ];
+		return migrateTextAlign( ...migrateToQuoteV2( attributes ) );
 	},
 };
 
@@ -195,9 +208,7 @@ const v2 = {
 		},
 	},
 	migrate( attributes ) {
-		const [ v2Attributes, innerBlocks ] = migrateToQuoteV2( attributes );
-		const v4Attributes = migrateTextAlign( v2Attributes );
-		return [ v4Attributes, innerBlocks ];
+		return migrateTextAlign( ...migrateToQuoteV2( attributes ) );
 	},
 	save( { attributes } ) {
 		const { align, value, citation } = attributes;
@@ -240,22 +251,12 @@ const v1 = {
 	migrate( attributes ) {
 		if ( attributes.style === 2 ) {
 			const { style, ...restAttributes } = attributes;
-			const [ v1Attributes, innerBlocks ] =
-				migrateToQuoteV2( restAttributes );
-			const v4Attributes = migrateTextAlign( v1Attributes );
-
-			return {
-				v4Attributes,
-				innerBlocks,
-				className: attributes.className
-					? attributes.className + ' is-style-large'
-					: 'is-style-large',
-			};
+			return migrateTextAlign(
+				...migrateLargeStyle( ...migrateToQuoteV2( restAttributes ) )
+			);
 		}
 
-		const [ v1Attributes, innerBlocks ] = migrateToQuoteV2( attributes );
-		const v4Attributes = migrateTextAlign( v1Attributes );
-		return [ v4Attributes, innerBlocks ];
+		return migrateTextAlign( ...migrateToQuoteV2( attributes ) );
 	},
 
 	save( { attributes } ) {
@@ -302,15 +303,10 @@ const v0 = {
 	migrate( attributes ) {
 		if ( ! isNaN( parseInt( attributes.style ) ) ) {
 			const { style, ...restAttributes } = attributes;
-			const [ v0Attributes, innerBlocks ] =
-				migrateToQuoteV2( restAttributes );
-			const v4Attributes = migrateTextAlign( v0Attributes );
-			return [ v4Attributes, innerBlocks ];
+			return migrateTextAlign( ...migrateToQuoteV2( restAttributes ) );
 		}
 
-		const [ v0Attributes, innerBlocks ] = migrateToQuoteV2( attributes );
-		const v4Attributes = migrateTextAlign( v0Attributes );
-		return [ v4Attributes, innerBlocks ];
+		return migrateTextAlign( ...migrateToQuoteV2( attributes ) );
 	},
 
 	save( { attributes } ) {
diff --git a/packages/block-library/src/quote/edit.js b/packages/block-library/src/quote/edit.js
index 44370ee44d..4689899c6a 100644
--- a/packages/block-library/src/quote/edit.js
+++ b/packages/block-library/src/quote/edit.js
@@ -132,9 +132,7 @@ export default function QuoteEdit( {
 								createBlock( getDefaultBlockName() )
 							)
 						}
-						{ ...( ! isWebPlatform
-							? { textAlign: attributes.textAlign }
-							: {} ) }
+						{ ...( ! isWebPlatform ? { textAlign } : {} ) }
 					/>
 				) }
 			</BlockQuotation>

```

</details>

carolinan commented:
I was not able to refactor migrateTextAlign as you suggested in August, so I am not sure I will be able to do this, but I will look at your suggested code changes.

aaronrobertshaw commented:
> I was not able to refactor migrateTextAlign as you suggested in August, so I am not sure I will be able to do this, but I will look at your suggested code changes.

FWIW I tested those suggested updates before adding them to the review. All the deprecation fixture tests passed locally here.

Update: Just retested and tests still pass for me. The steps I took were:
1. Copy the diff from the review
2. `pbpaste | git apply`
3. `npm run test:unit test/integration/full-content/full-content.test.js`

<img width="974" alt="Screenshot 2023-10-09 at 6 54 58 pm" src="https://github.com/WordPress/gutenberg/assets/60436221/8ad2407f-167c-483e-af5b-3bf02b503ca5">

carolinan committed:
Update migration, apply changes requested during review.

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

aaronrobertshaw reviewed (approved):
Nice work seeing this one across the line @carolinan 👍 

✅ Tests as advertised
✅ Fixture tests and deprecations LGTM

From the web side of things I think this is ready.  It also appears from this [comment](https://github.com/WordPress/gutenberg/pull/42960#pullrequestreview-1661938608) to be ready from the mobile side too but you might want to reconfirm with @SiobhyB.

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute

SiobhyB commented:
@carolinan, I'm just starting work on reviewing now. :)

SiobhyB reviewed (approved):
Approved from the native side! I manually tested to verify that the changes don't break the quote block in the app. I also created a Gutenberg Mobile PR to verify all automated tests pass for mobile here: https://github.com/wordpress-mobile/gutenberg-mobile/pull/6640.

Thanks for looping in the native team here @carolinan!

SiobhyB commented on packages/block-editor/src/components/block-list/block-list-item.native.js:
```diff
@@ -98,7 +98,7 @@ function BlockListItemContent( {
 			const parentName = getBlockName( rootClientId );
 			const { align } = getBlockAttributes( clientId ) || {};
-			const { align: parentBlockAlign } =
+			const { textAlign: parentBlockAlign } =
```
@carolinan, thinking on this more, I realised we may need to co-ordinate releases to ensure centred quote blocks don't break for app users. Would you be able to hold off from merging until I co-ordinate a release plan with the mobile team? I'll update here when we have something in place. Thank you!

carolinan replied:
Of course.

SiobhyB replied:
@carolinan, thank you for your patience here! The earliest app release we'd be able to get this into is `24.4`, which is set to be released to users on Monday, 18th March. With that in mind, could you merge this _after_ the `17.8` Gutenberg release is cut this coming Wednesday? If this is included in the `17.9` release (set to be released on March 13th), it'd minimise the number of days it's not available for app users. 

From my testing, I confirmed that having the change in place on the web but not in the app won't break the site’s content. The only issue I found is that the app's in-editor preview won’t reflect whether a Quote block is centred or right-aligned, which might be confusing. However, this will only be an issue for blocks that are first saved on the web then opened in the app. The in-app alignment buttons will continue to work and correct alignment will still be visible on the site’s front-end.

With the above in mind, we think minimising the number of days that the change is available on one platform but not another is the best option. Let me know if you're okay with delaying merge until after this Wednesday, for the `17.9` release. Thanks!

carolinan replied:
Yes, it can wait.

SiobhyB replied:
@carolinan, just to follow up, if you're able to merge this before this coming Thursday, we'll  then work to get it into the app's `24.4` release. Thanks again for all your help here!

carolinan committed:
Merge branch 'trunk' into update/rename-align-attribute
